File: app/src/main/java/com/wen/android/mtabuscomparison/ui/stopmap/StopMapFragment.java
package com.wen.android.mtabuscomparison.ui.stopmap;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.location.Location;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.model.LatLng;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.wen.android.mtabuscomparison.R;
import com.wen.android.mtabuscomparison.common.permission.MyPermission;
import com.wen.android.mtabuscomparison.common.permission.PermissionHelper;
import com.wen.android.mtabuscomparison.feature.stop.BusDatabase;
import com.wen.android.mtabuscomparison.feature.stop.Stop;
import com.wen.android.mtabuscomparison.feature.stop.StopInfo;
import com.wen.android.mtabuscomparison.ui.routesview.RoutesViewActivity;
import com.wen.android.mtabuscomparison.ui.search.SearchActivity;
import com.wen.android.mtabuscomparison.ui.stopmonitoring.StopMonitoringActivity;
import com.wen.android.mtabuscomparison.util.SearchHandler;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executors;

import timber.log.Timber;

/**
 * Created by yuan on 4/10/2017.
 */

public class StopMapFragment extends Fragment
        implements StopMapViewMvc.OnMovedMapListener,
        StopMapViewMvc.Listener,
        StopMapViewMvc.OnStartSearchListener,
        StopMapViewMvc.MapListener, PermissionHelper.Listener {

    private final int PERMISSION_ACCESS_FINE_LOCATION = 1;
    private FusedLocationProviderClient fusedLocationClient;
    public static final int SEARCH_ACTIVITY_REQUEST_CODE = 199;
    private PermissionHelper mPermissionHelper;
    private StopMapViewMvc mStopMapView;

    public StopMapFragment() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mPermissionHelper = new PermissionHelper(getActivity());
        if (!mPermissionHelper.hasPermission(MyPermission.FINE_LOCATION)) {
            requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                    PERMISSION_ACCESS_FINE_LOCATION);
        } else {
            fusedLocationClient = LocationServices.getFusedLocationProviderClient(getContext());
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mStopMapView = new StopMapViewMvcImpl(inflater, container, getChildFragmentManager(), this, this);

        setHasOptionsMenu(true);

        mStopMapView.getSearchBar().setOnEditorActionListener((v, actionId, event) -> {
            displaySearchResult(v.getText().toString());
            return true;
        });
        return mStopMapView.getRootView();
    }

    @Override
    public void onResume() {
        mStopMapView.onResume();
        super.onResume();
    }

    @Override
    public void onStart() {
        super.onStart();
        mStopMapView.registerMapListener(this);
        mStopMapView.registerListener(this);
        mPermissionHelper.registerListener(this);
    }

    @Override
    public void onPause() {
        mStopMapView.onPause();
        super.onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        mStopMapView.unregisterMapListener(this);
        mStopMapView.unregisterListener(this);
        mPermissionHelper.unregisterListener(this);
    }

    @Override
    public void onDestroyView() {
        mStopMapView.onDestroy();
        super.onDestroyView();
    }

    /**
     * start a new activity and display the search result
     */
    public void displaySearchResult(String userInput) {
        Bundle bundle = new Bundle();
        bundle.putString(FirebaseAnalytics.Param.SEARCH_TERM, userInput);
        FirebaseAnalytics.getInstance(getContext()).logEvent(FirebaseAnalytics.Event.SEARCH, bundle);
        SearchHandler searchHandler = new SearchHandler(userInput);
        if (searchHandler.keywordType() == 0) {
            String[] stopcodeArray = new String[1];
            //get the bus code from the user input
            stopcodeArray[0] = userInput;
            if (stopcodeArray[0] == null) {
                return;
            }
            Intent intent = new Intent(getActivity(), StopMonitoringActivity.class);
            intent.putExtra(Intent.EXTRA_TEXT, stopcodeArray);
            startActivity(intent);
        } else {
            String routeEntered = userInput.toUpperCase();
            Intent intent = new Intent(getActivity(), RoutesViewActivity.class);
            intent.putExtra(Intent.EXTRA_TEXT, routeEntered);
            startActivity(intent);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        mPermissionHelper.onRequestPermissionResult(requestCode, permissions, grantResults);
    }

    /**
     * get the current location and then find nearby stop from database
     */
    private void findNearByStop(Location location) {
        final ArrayList<StopInfo> stopList = new ArrayList<>();
        //change radius_in_meters if we want to change the range of the nearby stop
        double radius_in_meters = 800;
        double radius_pls = radius_in_meters;
        double radius_neg = 0 - radius_in_meters;
        double coef_plus = radius_pls * 0.0000089;
        double coef_neg = radius_neg * 0.0000089;
        double new_latitude1, new_latitude2, new_longitude1, new_longitude2;

        double current_latitude = location.getLatitude();
        double current_longitude = location.getLongitude();

        new_latitude1 = current_latitude + coef_neg;
        new_latitude2 = current_latitude + coef_plus;

        new_longitude1 = current_longitude + coef_plus / Math.cos(current_latitude * 0.018);
        new_longitude2 = current_longitude + coef_neg / Math.cos(current_latitude * 0.018);

        Executors.newSingleThreadExecutor().execute(() -> {
            List<Stop> bustList = BusDatabase.Companion.getInstance(getContext()).allBusDao().getStopsInRange(new_latitude1, new_latitude2, new_longitude2, new_longitude1);
            for (Stop bus : bustList) {
                StopInfo stop = new StopInfo();
                Location tempLocation = new Location("tempLocation");
                tempLocation.setLatitude(bus.getStopLat());
                tempLocation.setLongitude(bus.getStopLon());

                float distance = location.distanceTo(tempLocation);
                stop.setStopCode(bus.getStopId());
                stop.setIntersections(bus.getStopName());
                stop.setRoutes(bus.getRouteId());
                stop.setLocation(tempLocation);
                stop.setDistance(distance);
                stopList.add(stop);
                Collections.sort(stopList);
            }
            if (getActivity() != null) {
                getActivity().runOnUiThread(() -> {
                    mStopMapView.removeMarkers();
                    for (StopInfo st : stopList) {
                        mStopMapView.addStopMarker(st);
                    }

                    updateNearbyStopList(stopList);
                });
            }
        });

    }

    private void updateNearbyStopList(List<StopInfo> nearbyStopList) {
        mStopMapView.bindStopInfo(nearbyStopList);
    }

    @Override
    public void onMovedMap(LatLng latLng) {
        Location location = new Location("");
        location.setLatitude(latLng.latitude);
        location.setLongitude(latLng.longitude);
        findNearByStop(location);
    }

    @Override
    public void onStopClicked(@NotNull StopInfo stop) {
        String[] stopcodeArray = new String[1];
        stopcodeArray[0] = stop.getStopCode();
        Intent intent = new Intent(getContext(), StopMonitoringActivity.class);
        intent.putExtra(Intent.EXTRA_TEXT, stopcodeArray);
        startActivity(intent);
    }

    @Override
    public void onStartSearch() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            startActivityForResult(
                    new Intent(getContext(), SearchActivity.class),
                    SEARCH_ACTIVITY_REQUEST_CODE,
                    ActivityOptions.makeSceneTransitionAnimation(getActivity()).toBundle()
            );
        } else {
            startActivityForResult(
                    new Intent(getContext(), SearchActivity.class),
                    SEARCH_ACTIVITY_REQUEST_CODE
            );
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == SEARCH_ACTIVITY_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                LatLng latlng = (LatLng) data.getParcelableExtra(getString(R.string.SEARCH_RESULT_POINT));
                mStopMapView.moveCameraTo(latlng);
            }

        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }

    }

    @SuppressLint("MissingPermission")
    @Override
    public void onMapReady() {
        if (mPermissionHelper.hasPermission(MyPermission.FINE_LOCATION)) {
            fusedLocationClient.getLastLocation().addOnSuccessListener(location -> {
                Timber.i("last known location: " + location);
                if (location == null) {
                    location = new Location("dummy");
                    location.setLatitude(40.74859491079061);
                    location.setLongitude(-73.98564294403914);
                }
                mStopMapView.addCurrentLocationMarker(location);
                findNearByStop(location);
            });
        }
    }

    @SuppressLint("MissingPermission")
    @Override
    public void onRequestPermissionsResult(int requestCode, @NotNull PermissionHelper.PermissionsResult result) {
        switch (requestCode) {
            case PERMISSION_ACCESS_FINE_LOCATION:
                if (result.getGranted() != null && result.getGranted().size() > 0 && result.getGranted().contains(MyPermission.FINE_LOCATION)) {
                    Timber.i("permission granted");
                    fusedLocationClient = LocationServices.getFusedLocationProviderClient(getContext());
                    fusedLocationClient.getLastLocation().addOnSuccessListener(location -> {
                        if (location == null) {
                            Timber.i("dummy??");
                            location = new Location("dummy");
                            location.setLatitude(40.74859491079061);
                            location.setLongitude(-73.98564294403914);
                        } else {
                            Timber.i("not dummy!" + location.toString());
                        }
                        mStopMapView.enableMyLocationButton();
                        findNearByStop(location);
                        mStopMapView.addCurrentLocationMarker(location);
                    });
                } else {
                    Toast.makeText(getContext(), "Need Location permission", Toast.LENGTH_LONG)
                            .show();
                }
                break;
        }
    }

    @Override
    public void onPermissionsRequestCancelled(int requestCode) {
    }
}

File: app/src/main/java/com/wen/android/mtabuscomparison/ui/stopmap/StopMapFragment.kt
package com.wen.android.mtabuscomparison.ui.stopmap

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.Bundle
import android.os.Handler
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.view.doOnLayout
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.NavHostFragment
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.maps.android.ktx.awaitMap
import com.wen.android.mtabuscomparison.R
import com.wen.android.mtabuscomparison.common.permission.MyPermission
import com.wen.android.mtabuscomparison.common.permission.PermissionHelper
import com.wen.android.mtabuscomparison.common.permission.PermissionHelper.PermissionsResult
import com.wen.android.mtabuscomparison.databinding.FragmentStopMapBinding
import com.wen.android.mtabuscomparison.feature.stopmonitoring.StopInfo
import com.wen.android.mtabuscomparison.util.bitmapDescriptorFromVector
import com.wen.android.mtabuscomparison.util.dpToPx
import com.wen.android.mtabuscomparison.util.fragment.repeatOnViewLifecycle
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by yuan on 4/10/2017.
 */
@AndroidEntryPoint
class StopMapFragment :
    Fragment(),
    PermissionHelper.Listener,
    StopsRecyclerAdapter.Listener {

    @Inject
    lateinit var permissionHelper: PermissionHelper

    @Inject
    lateinit var firebaseAnalytics: FirebaseAnalytics

    private var currentFocusStop = 0
    private var previousFocusStop = Integer.MAX_VALUE

    //    private val stopMarkerList = mutableListOf<Marker>()
    private val noUpdateDistance = 200
    private var previousCameraLocation = Location("").apply {
        this.latitude = 0.0
        this.longitude = 0.0
    }

    private lateinit var stopAdapter: StopsRecyclerAdapter

    private val viewModel: StopMapViewModel by viewModels()

    private var myLocationJob: Job? = null
    private var myLocationOnPermissionJob: Job? = null

    private var _binding: FragmentStopMapBinding? = null
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Timber.v("onCreate")

        /**
         * request permission
         * nearby stop is the only functionality that needs location permission now
         * if permission is rejected, just disable nearby stop
         */
        if (!permissionHelper.hasPermission(MyPermission.FINE_LOCATION)) {
            if (shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_FINE_LOCATION)) {
                MyLocationRationaleFragment()
                    .show(childFragmentManager, FRAGMENT_MY_LOCATION_RATIONAL)
                return
            }
            requestPermissions(
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                PERMISSION_ACCESS_FINE_LOCATION
            )
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        Timber.v("onCreateView")
        _binding = FragmentStopMapBinding.inflate(inflater, container, false).apply {
            viewModel = this@StopMapFragment.viewModel
        }

        binding.stopMapMapView.apply {
            onCreate(savedInstanceState)
        }

        binding.apply {
            searchView.apply {
                doOnLayout {
                    it.translationX = it.width - (40.dpToPx).toFloat()
                }
            }
            stopMapSearchIcon.setOnClickListener { onStartSearch() }
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        NavHostFragment.findNavController(this).apply {
            currentBackStackEntry?.savedStateHandle?.getLiveData<LatLng>(getString(R.string.SEARCH_RESULT_POINT))
                ?.observe(viewLifecycleOwner) { latLng ->
                    Timber.i("result from search $latLng")
                    currentBackStackEntry?.savedStateHandle?.remove<LatLng>(getString(R.string.SEARCH_RESULT_POINT))
                    Handler().postDelayed(Runnable {
                        moveCameraTo(latLng)
                    }, 300)
                }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            binding.stopMapMapView.awaitMap().apply {

                setOnMarkerClickListener { marker ->
                    val stopList = stopAdapter.mStops
                    for (position in stopList.indices) {
                        val lat = marker.position.latitude.compareTo(stopList[position].location.latitude)
                        val lon = marker.position.longitude.compareTo(stopList[position].location.longitude)
                        if (lat == 0 && lon == 0) {
                            previousFocusStop = currentFocusStop
                            currentFocusStop = position
                            scrollToStop(position)
                        }
                    }
                    false
                }
                enableMyLocationButton(this)
                onMapReady(this)
                setOnCameraIdleListener {
                    val currentCameraLocation = Location("").also {
                        it.latitude = this.cameraPosition.target.latitude
                        it.longitude = this.cameraPosition.target.longitude
                    }
                    if (previousCameraLocation.distanceTo(currentCameraLocation) > noUpdateDistance) {
                        onMovedMap(this.cameraPosition.target)
                    }
                    previousCameraLocation = currentCameraLocation
                }
            }
        }

        repeatOnViewLifecycle {
            viewModel.nearByStop.collect {
                withContext(Dispatchers.Main) {
                    removeMarkers()
                    for (stopInfo in it) {
                        addStopMarker(stopInfo)
                    }
                    updateNearbyStopList(it)
                }

            }
        }
    }

    override fun onResume() {
        Timber.i("onResume")
        super.onResume()
        binding.stopMapMapView.onResume()

        firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, StopMapFragment::class.java.simpleName)
        }
    }

    override fun onStart() {
        Timber.i("onStart")
        super.onStart()
        binding.stopMapMapView.onStart()
        permissionHelper.registerListener(this)
    }

    override fun onPause() {
        Timber.i("onPause")
        super.onPause()
        binding.stopMapMapView.onPause()
    }

    override fun onStop() {
        Timber.i("onStop")
        super.onStop()
        myLocationJob?.cancel()
        myLocationOnPermissionJob?.cancel()
        binding.stopMapMapView.onStop()
        permissionHelper.unregisterListener(this)
    }

    override fun onDestroyView() {
        Timber.v("onDestroyView()")
        binding.nearbyRecycleView.adapter = null
        binding.stopMapMapView.onDestroy()
        binding.stopMapMapView.removeAllViews()
        _binding = null
        super.onDestroyView()
    }

    override fun onDestroy() {
        Timber.v("onDestroy()")
        super.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        binding.stopMapMapView.onLowMemory()
    }

    private fun bindStopInfo(data: List<StopInfo>) {
        stopAdapter = StopsRecyclerAdapter(data, this, this)
        currentFocusStop = 0
        binding.nearbyRecycleView.adapter = stopAdapter
    }

    private fun scrollToStop(index: Int) {
        binding.nearbyRecycleView.layoutManager!!.scrollToPosition(index)
        binding.nearbyRecycleView.adapter!!.notifyItemChanged(previousFocusStop)
        binding.nearbyRecycleView.adapter!!.notifyItemChanged(index)
    }

    private fun removeMarkers() {
        viewLifecycleOwner.lifecycleScope.launch {
            val map = binding.stopMapMapView.awaitMap()
            map.clear()
        }
    }

    private fun addStopMarker(st: StopInfo) {
        val nearbyStopLatLng = LatLng(st.location.latitude, st.location.longitude)
        viewLifecycleOwner.lifecycleScope.launch {
            val map = binding.stopMapMapView.awaitMap()
            map.addMarker(
                MarkerOptions()
                    .position(nearbyStopLatLng)
                    .icon(bitmapDescriptorFromVector(requireContext(), R.drawable.ic_bus_blue_20dp))
                    .title(st.intersections)
            )
        }
    }

    private fun moveCameraTo(latLng: LatLng) {
        viewLifecycleOwner.lifecycleScope.launch {
            val map = binding.stopMapMapView.awaitMap()
            Timber.i("move camera to $latLng")
            map.animateCamera(
                CameraUpdateFactory.newCameraPosition(
                    CameraPosition.builder()
                        .target(latLng)
                        .zoom(17f)
                        .build()
                )
            )
        }

    }

    fun getFocusStop() = currentFocusStop

    @SuppressLint("ResourceType")
    fun enableMyLocationButton(map: GoogleMap) {
        if (ActivityCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            map.isMyLocationEnabled = true
            binding.stopMapMapView.findViewById<View>(
                0x2
            ).apply {
                val params = layoutParams as RelativeLayout.LayoutParams
                params.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM)
                params.addRule(RelativeLayout.ALIGN_PARENT_LEFT)
                params.addRule(RelativeLayout.ALIGN_PARENT_RIGHT, 0)
                params.addRule(RelativeLayout.ALIGN_PARENT_TOP, 0)
                params.bottomMargin = 50.dpToPx
                params.rightMargin = 0.dpToPx
                layoutParams = params
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        permissionHelper.onRequestPermissionResult(requestCode, permissions, grantResults)
    }

    private fun updateNearbyStopList(nearbyStopList: List<StopInfo>) {
        bindStopInfo(nearbyStopList)
    }

    private fun onMovedMap(latLng: LatLng) {
        val location = Location("")
        location.latitude = latLng.latitude
        location.longitude = latLng.longitude
        viewModel.loadNearByStop(location)
    }

    override fun onStopClicked(stop: StopInfo) {
        val navController = NavHostFragment.findNavController(this)
        navController.navigate(
            StopMapFragmentDirections.actionStopMapFragmentToStopMonitoringFragment(stop.stopCode)
        )
    }

    private fun onStartSearch() {
        NavHostFragment.findNavController(this).apply {
            navigate(StopMapFragmentDirections.actionStopMapFragmentToSearchFragment())
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == SEARCH_ACTIVITY_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                if (data!!.hasExtra(getString(R.string.SEARCH_RESULT_STOP_CODE))) {
                    goToStopMonitoringFragment(data.getStringExtra(getString(R.string.SEARCH_RESULT_STOP_CODE))!!)
                } else {
                    val latlng = data.getParcelableExtra<Parcelable>(getString(R.string.SEARCH_RESULT_POINT)) as LatLng
                    moveCameraTo(latlng)
                }
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    private fun goToStopMonitoringFragment(stopCode: String) {
        val navController = NavHostFragment.findNavController(this)
        navController.navigate(
            StopMapFragmentDirections.actionStopMapFragmentToStopMonitoringFragment(stopCode)
        )
    }

    @SuppressLint("MissingPermission")
    private fun onMapReady(map: GoogleMap) {
        Timber.v("onMapReady()")
        if (permissionHelper.hasPermission(MyPermission.FINE_LOCATION)) {
            /**
             * one shot update
             * doesn't want to repeat this after onStop
             *
             */
            myLocationJob = viewLifecycleOwner.lifecycleScope.launchWhenStarted {
                viewModel.myLocationCameraUpdate
                    .collect {
                        it?.let {
                            Timber.v("collecting from myLocationCameraUpdate, onMapReady()")
                            map.moveCamera(it)
                        }
                    }
            }
        }
    }

    @SuppressLint("MissingPermission")
    override fun onRequestPermissionsResult(requestCode: Int, result: PermissionsResult) {
        when (requestCode) {
            PERMISSION_ACCESS_FINE_LOCATION -> if (result.granted != null
                && result.granted.isNotEmpty()
                && result.granted.contains(
                    MyPermission.FINE_LOCATION
                )
            ) {
                Timber.i("permission granted")

                /**
                 * one shot update
                 * doesn't want to repeat this after onStop
                 */
                myLocationOnPermissionJob = viewLifecycleOwner.lifecycleScope.launchWhenStarted {
                    binding.stopMapMapView.awaitMap().apply {
                        enableMyLocationButton(this)
                        viewModel.myLocationCameraUpdate
                            .collect {
                                it?.let {
                                    Timber.v("collecting from myLocationCameraUpdate")
                                    this.moveCamera(it)
                                }
                            }
                    }

                }

            } else {
                Toast.makeText(context, "Need Location permission", Toast.LENGTH_LONG)
                    .show()
            }
        }
    }

    override fun onPermissionsRequestCancelled(requestCode: Int) {}

    companion object {
        const val SEARCH_ACTIVITY_REQUEST_CODE = 199
        private const val PERMISSION_ACCESS_FINE_LOCATION = 1
        private const val FRAGMENT_MY_LOCATION_RATIONAL = "my_location_rational"
    }

    class MyLocationRationaleFragment : DialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            return MaterialAlertDialogBuilder(requireContext())
                .setMessage(R.string.my_location_rational)
                .setPositiveButton(android.R.string.ok) { _, _ ->
                    requireParentFragment().requestPermissions(
                        arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                        PERMISSION_ACCESS_FINE_LOCATION
                    )
                }
                .setNegativeButton(android.R.string.cancel, null)
                .create()
        }
    }
}